File: cine/solutions/correct/nlehmann.cpp
#include <algorithm>
#include <cstdio>
#include <iostream>
#include <vector>
using namespace std;

int N, M, K;

bool try_radius(vector<vector<int>> &sums, vector<vector<int>> &seats, int r) {
  // Probar sentar a Pedro en todas las posiciones
  for (int y = 1; y <= N; ++y) {
    for (int x = 1; x <= M; ++x) {
      // Asiento ocupado
      if (seats[y][x]) {
        continue;
      }

      // Contar la cantidad de ceros en la submatriz de radio `r` centrada en
      // `(x, y)` asegurandose que esté totalmente contenida.
      int y1 = max(y - r, 1), y2 = min(y + r, N);
      int x1 = max(x - r, 1), x2 = min(x + r, M);
      int c = sums[y2][x2] - sums[y1 - 1][x2] - sums[y2][x1 - 1] +
              sums[y1 - 1][x1 - 1];

      // Si caben todos respondemos `true`
      if (c >= K + 1) {
        return true;
      }
    }
  }
  return false;
}

using namespace std;

int main() {
  cin >> N >> M >> K;

  // Los asientos
  vector<vector<int>> seats(N + 1, vector<int>(M + 1));
  // Suma de prefijos para calcular la cantidad de ceros en una submatriz
  vector<vector<int>> sums(N + 1, vector<int>(M + 1, 0));
  for (int y = 1; y <= N; ++y) {
    for (int x = 1; x <= M; ++x) {
      cin >> seats[y][x];
      sums[y][x] =
          !seats[y][x] + sums[y - 1][x] + sums[y][x - 1] - sums[y - 1][x - 1];
    }
  }

  // Busqueda binaria en el "radio", es decir, busqueda binaria en las posibles
  // distancias grupales. `max(N, M)` es una cota superior olgada. Podría
  // optimizarse y solo buscar hasta la mitad.
  int left = 1;
  int right = max(N, M);
  while (left < right) {
    int mid = left + (right - left) / 2;
    if (try_radius(sums, seats, mid)) {
      right = mid;
    } else {
      left = mid + 1;
    }
  }
  cout << left << endl;

  return 0;
}

File: cine/solutions/correct/nlehmann.java
import java.util.*;
import java.io.*;

class cine {
    static public void main(String[] args) throws Exception {
        var in = new BufferedReader(new InputStreamReader(System.in));
        var st = new StringTokenizer(in.readLine());

        int N = Integer.parseInt(st.nextToken());
        int M = Integer.parseInt(st.nextToken());
        int K = Integer.parseInt(st.nextToken());

        var seats = new int[N + 1][M + 1];
        var sums = new int[N + 1][M + 1];
        for (int y = 1; y <= N; ++y) {
            st = new StringTokenizer(in.readLine());
            for (int x = 1; x <= M; ++x) {
                seats[y][x] = Integer.parseInt(st.nextToken());
                sums[y][x] = (seats[y][x] == 0 ? 1 : 0) + sums[y-1][x] + sums[y][x-1] - sums[y-1][x-1];
            }
        }

        int min = Math.max(N, M);
        for (int y = 1; y <= N; ++y) {
            for (int x = 1; x <= M; ++x) {
                if (seats[y][x] == 1) {
                    continue;
                }
                int left = 1;
                int right = Math.max(N, M);
                while (left < right) {
                    int mid = left + (right - left)/2;

                    int x1 = Math.max(x-mid, 1), x2 = Math.min(x+mid, M);
                    int y1 = Math.max(y-mid, 1), y2 = Math.min(y+mid, N);

                    int count = sums[y2][x2] - sums[y2][x1-1] - sums[y1-1][x2] + sums[y1-1][x1-1];

                    if (count >= K + 1) {
                        right = mid;
                    } else {
                        left = mid + 1;
                    }
                }
                min = Math.min(left, min);
            }
        }
        System.out.println(min);
    }

}
